compiler/optimizer/OMRSimplifierHandlers.cpp

```diff
@@ -12647,6 +12659,195 @@ TR::Node *su2dSimplifier(TR::Node * node, TR::Block * block, TR::Simplifier * s)
+            unsignedMax = isUnsigned  ? TR::getMaxUnsigned<TR::Int16>() : 0;
+            unsignedMin = isUnsigned  ? TR::getMinUnsigned<TR::Int16>() : 0;
+            }
+         else if (firstChild->getOpCode().is4Byte())
+            {
+            signedMax   = !isUnsigned ? TR::getMaxSigned<TR::Int32>()   : 0;
+            signedMin   = !isUnsigned ? TR::getMinSigned<TR::Int32>()   : 0;
+            unsignedMax = isUnsigned  ? TR::getMaxUnsigned<TR::Int32>() : 0;
+            unsignedMin = isUnsigned  ? TR::getMinUnsigned<TR::Int32>() : 0;
+            }
+         else
+            {
+            TR_ASSERT(0, "Unable to handle a number that's bigger than 4 bytes.");
+            }
+
+         if ((isUnsigned && (const3 > unsignedMax || const3 < unsignedMin)) ||
+                 (!isUnsigned && (const3 > signedMax || const3 < signedMin)))
+            {
+            return node;
+            }
+         }
+      else  // Long
+         {
+         if (isUnsigned)
+            {
+            uint64_t uconst1 = firstChild->getSecondChild()->getUnsignedLongInt();
+            uint64_t uconst2 = secondChild->getUnsignedLongInt();
+            uint64_t uconst3 = 0;
+
+            if (isAdd && !IS_UINT64_SUB_UNDERFLOW(const2, const1))
+               {
+               uconst3 = uconst2 - uconst1;
+               }
+            else if (isSub && !IS_UINT64_ADD_OVERFLOW(const2, const1))
+               {
+               uconst3 = uconst2 + uconst1;
+               }
+            else
+               {
+               return node;
+               }
+            }
+         else
+            {
+            const1 = firstChild->getSecondChild()->getConstValue();
+            const2 = secondChild->getConstValue();
+
+            if (isAdd &&
+                    !IS_INT64_SUB_OVERFLOW(const2, const1) &&
+                    !IS_INT64_SUB_UNDERFLOW(const2, const1))
+               {
+               const3 = const2 - const1;
+               }
+            else if (isSub &&
+                     !IS_INT64_ADD_OVERFLOW(const2, const1) &&
+                     !IS_INT64_ADD_UNDERFLOW(const2, const1))
+               {
+               const3 = const2 + const1;
+               }
+            else
+               {
+               return node;
+               }
+            }
+         }
+
+      // Done checking constant values. Change the nodes.
+      dumpOptDetails(s->comp(), "%sEliminating add/sub operation under comparison node n%dn\n",
+                     s->optDetailString(),
+                     node->getGlobalIndex());
+
+      // Handle the iconst node in the compare
+      bool isUnsignedLong = isUnsigned && !isInt;
+      if (secondChild->isSingleRef())
+         {
+         if (isUnsignedLong)
+            {
+            secondChild->setUnsignedLongInt(const3);
+            }
+         else
+            {
+            secondChild->setConstValue(const3);
+            }
+         }
```

Comment:
Historically there has been code that changes constant values in-place, but it causes issues. For example, from time to time I've seen constants with `(X>=0 X<=0 X!=0)`.

The method is still useful for creating a new constant node in this case, since there is currently no `Node::create` for integer constants of arbitrary size. There is a variant that takes `int32_t intValue`, but it claims to be invalid for smaller integer types, and it's unsuitable for long constants.

We also need to set the value when transmuting a previously non-constant node into a constant. For example, with

    iadd
      iconst 3
      iconst 4

We do something like the following, which does not change the node's value:

    iaddNode->removeAllChildren();
    TR::Node::recreate(iaddNode, TR::iconst);
    iaddNode->setInt(7); // computed from the children, of course

It could have been set by a variant of `recreate` if we had one, and that might be an improvement, but having access to this method is no worse in this sense than having access to `setAndIncChild()`.